### PR TITLE
Use PSPBUILDs for building libraries

### PR DESCRIPTION
**Pull request by sharkwouter**
With this PR the psplibraries repo will be replaced with a set of PSPBUILDs and probably a build script. Installing within  the toolchain should be changed to use a repository when this is merged.

This is still work in progress. The following libraries need to be ported to PSPBUILDs:

- [x] angelscript
- [x] bzip2
- [x] psp-cmake ~~(Still needs license)~~
- [x] expat
- [x] freetype (switched away from psp-ports and renamed to freetype2)
- [x] jpeg (requires psp-ports)
- [x] libbulletml (requires psp-ports)
- [x] libmad (requires psp-ports)
- [x] libmikmod
- [x] libogg (requires psp-ports)
- [x] libpng
- [x] libpspvram
- [x] libtremor (requires psp-ports)
- [x] libvorbis (requires psp-ports)
- [x] libyaml
- [x] lua (requires psp-ports)
- [x] openal (renamed to openal-soft)
- [x] opentri
- [x] pixman
- [x] pspgl (requires psp-ports)
- [x] pspirkeyb (requires psp-ports)
- [x] pthreads-emb (renamed to pthread-embedded)
- [x] SDL
- [x] SDL2
- [x] SDL2_gfx (requires psp-ports)
- [x] SDL2_image (requires psp-ports)
- [x] SDL2_mixer (requires psp-ports)
- [x] SDL2_ttf (requires psp-ports)
- [x] SDL_gfx (requires psp-ports)
- [x] SDL_image (requires psp-ports)
- [x] SDL_mixer (requires psp-ports)
- [x] SDL_ttf (requires psp-ports)
- [x] smpeg-psp
- [x] sqlite
- [x] squirrel
- [x] zlib
- [x] zziplib (requires psp-ports)

Currently I've removed the scripts which I won't be using at all and kept the scripts which can still serve as reference while porting this repo. These will also be removed later.

I've also added contributor guidelines which states the requirements a PSPBUILD should meet to be accepted into this repo.

I have not figured out yet how to build libraries which depend upon other libraries within this repo. There must be some way to do this automatically.

We might want to make a separate branch for this work, so I won't have to port 37 libraries on my own.

### CODE REVIEW

**Comment by sharkwouter** [comment]
For libpspvram, I'm using this repo as source instead of psp-ports: https://github.com/albe/libpspvram

File: bzip2/PSPBUILD [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+prepare() {
+	cd "$pkgname-$pkgver"
+	patch -p1 < ../${pkgname}-${pkgver}-PSP.patch
+}
+
+build() {
+	cd "$pkgname-$pkgver"
+	make -j`nproc`
```
Comment (carstene1ns) [inline bzip2/PSPBUILD:28]: This is wrong. You only ever want to set that to `-j1`, if the Makefile does not support parallel builds. This is set globally in `makepkg.conf`.

File: libpspvram/PSPBUILD [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+license=('MIT')
+depends=()
+makedepends=()
+optdepends=()
+source=("git+https://github.com/albe/libpspvram.git")
+sha256sums=('SKIP')
+
+pkgver() {
+  cd "$pkgname"
+  # Set the revisioin of the repo as version
```
Comment (carstene1ns) [inline libpspvram/PSPBUILD:17]: typo.

File: libpspvram/PSPBUILD [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+build() {
+	cd "$pkgname"
+	make
+}
+
+package() {
+	cd "$pkgname"
+	# make install isn't usable for pspgl, so we're installing manually
```
Comment (carstene1ns) [inline libpspvram/PSPBUILD:28]: leftover.

File: libpspvram/PSPBUILD [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	install -Dm 644 "libpspvalloc.a" "$pkgdir/psp/lib/libpspvalloc.a"
+	install -Dm 644 "libpspvram.a" "$pkgdir/psp/lib/libpspvram.a"
+
+	mkdir -m 755 -p "$pkgdir/psp/include/"
+	install -Dm 644 *.h "$pkgdir/psp/include/"
```
Comment (carstene1ns) [inline libpspvram/PSPBUILD:33]: `D` option is wrong.

File: pspgl/PSPBUILD [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+depends=()
+makedepends=()
+optdepends=()
+source=("git+https://github.com/pspdev/psp-ports.git")
+sha256sums=('SKIP')
+
+pkgver() {
+  cd psp-ports
+  # Set the revisioin of the repo as version
+  # Means any update to psp-ports will result in a new version number
```
Comment (carstene1ns) [inline pspgl/PSPBUILD:18]: This versioning will not work out. As said we want to move all psp-ports out in separate repos.

File: pspgl/PSPBUILD [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  printf "r%s.%s" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"
+}
+
+prepare() {
+	cd "psp-ports/$pkgname"
+	for i in glLockArraysEXT pspgl_dlist pspgl_vidmem
+	do
+		echo '#include <psptypes.h>' > $i.c.tmp
```
Comment (carstene1ns) [inline pspgl/PSPBUILD:26]: This is ugly. Why not apply in the psp-port instead?

File: pspgl/PSPBUILD [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+		cat $i.c >> $i.c.tmp
+		cat $i.c.tmp > $i.c
+	done
+}
+
+build() {
+	cd "psp-ports/$pkgname"
+	make -j`nproc`
```
Comment (carstene1ns) [inline pspgl/PSPBUILD:34]: see above.

File: zlib/PSPBUILD [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+prepare() {
+	cd "$pkgname-$pkgver"
+	patch -p1 < ../${pkgname}-${pkgver}-PSP.patch
+}
+
+build() {
+	cd "$pkgname-$pkgver"
+	make -j`nproc`
```
Comment (carstene1ns) [inline zlib/PSPBUILD:28]: see above.

File: pspgl/PSPBUILD (hunk 5 above)
Comment (sharkwouter) [reply pspgl/PSPBUILD:18]: Yeah, agreed. This was probably the best solution for when using the psp-ports repo, but it's not a good one.

File: pspgl/PSPBUILD (hunk 6 above)
Comment (sharkwouter) [reply pspgl/PSPBUILD:26]: I was wondering this too, since it's in the build script in psp-ports. I can fix this later.

Comment (sharkwouter) [reply pspgl/PSPBUILD:26]: I've made the following PRs to address this:
- https://github.com/pspdev/psplibraries/pull/67
- https://github.com/pspdev/psp-ports/pull/33

When these have been accepted, this code should no longer be needed. Do merge both of them at the same time, though, otherwise there will be issues.

**Comment by sharkwouter** [comment]
We should probably define some groups. There should be a group for all the packages in this library probably, something like ``psplibraries``. Then you could install all of them with ``psp-pacman -S psplibraries``.

There could also be groups for SDL and SDL2, something like ``psplibraries-sdl`` could work. Dunno if any other groups would make sense.

**Comment by sharkwouter** [comment]
Thanks to @Sir-Fortesque sqlite has been added.

**Comment by sharkwouter** [comment]
I've added 2 script files called ``libraries.sh`` and ``parse_pspbuild.sh``. The ``libraries.sh`` script builds and installs all PSPBUILDs in this github repository and it takes the dependencies into account while doing so. A ``PSPBUILD`` will only be build when all dependencies have been build and installed. It will throw some warnings if a package has already been build in an earlier step, but this probably isn't an issue.

The ``parse_pspbuild.sh`` script exists because I use the ``source`` command to get the variables in a ``PSPBUILD``. This needed to be in a separate script to keep my environment clean inside ``libraries.sh``.

This implementation does make a couple of assumptions, though:
- All PSPBUILDs are found at ``./*/PSPBUILD``.
- The name of the package and the name of the directory it's in are always the same.
- We don't use ``provides`` in our PSPBUILDs.
- All PSPBUILDs in this git repository are valid.
- All dependencies found in PSPBUILDs have a ``PSPBUILD`` within this git repository.
- PSPBUILDs in this git repository do not execute code when importing them using the ``source`` command.

Le me know if any of these are problematic.

**Comment by sharkwouter** [comment]
@wally2000 and @dbeef what should the license on psp-cmake and the toolchain files be? I've packaged it now, but there is no license. The toolchain script and the createpbp script are long enough to need one.

**Comment by dbeef** [comment]
> @wally4000 and @dbeef what should the license on psp-cmake and the toolchain files be? I've packaged it now, but there is no license. The toolchain script and the createpbp script are long enough to need one.

I'd say MIT license for `CreatePBP.cmake`. I can't say for the cmake toolchain file, as I am not the only contributor.

**Comment by sharkwouter** [comment]
After chatting with dbeef and wally, it was decided that we'd use GPL for psp-cmake. I've added it.

File: pspgl/PSPBUILD (hunk 6 above)
Comment (sharkwouter) [reply pspgl/PSPBUILD:26]: The mentioned PRs have been merged and the code removed.

File: expat/PSPBUILD [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	make
+}
+
+package() {
+	cd "$pkgname-$pkgver"
+	make DESTDIR="$pkgdir/" install
+	
+	install -Dm 644 COPYING "$pkgdir/psp/share/licenses/$pkgname/COPYING"
+
+	# Remove the binary we can't use
```
Comment (carstene1ns) [inline expat/PSPBUILD:41]: Might want to disable with configure:
https://github.com/libexpat/libexpat/blob/master/expat/configure.ac#L128-L144
or build with cmake, needs version update though.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    - ``pkgdesc``
+    - ``license``
+  - ``arch`` should be set to ``('mips')`` for PSP libraries or to ``('any')`` for scripts.
+  - ``sha256sums`` should be used for integrity checks of downloaded files. Only version controlled repositories (think git, svn or mercurial) and local patches are allowed to use ``SKIP``.
+  - PSPBUILDs based on versioned archive files (yourlibrary-1.2.tar.gz for instance) are preferred over those based on version controlled repositories.
+  - The license of the library should be installed in ``$pkgdir/psp/share/licenses/$pkgname``.
+  - For PSPBUILDs based on a version controlled repository, use the ``pkgver()`` function to set the ``pkgver`` to something which only changes if that repository is updated. See [this page](https://wiki.archlinux.org/index.php/VCS_package_guidelines) for instructions on how to do that.
+  - No packages which can not be legally distributed after building are allowed.
+  - ``pkgname`` should not contain capital letters or special characters other than ``-``.
+  - PSPBUILDS should start with ``#Maintainer: Your Name``. Adding an email at the end like ``<youemail@example.com>`` is optional, but encouraged.
```
Comment (carstene1ns) [inline CONTRIBUTING.md:18]: I would suggest we leave maintainership in pspdev organisation, provide a template PSPBUILD and let
contributors add themselves as what they are: `# Contributor: random <r@nd.om>`

Comment (sharkwouter) [reply CONTRIBUTING.md]: That makes a lot of sense. I've changed it.

File: expat/PSPBUILD (hunk 9 above)
Comment (sharkwouter) [reply expat/PSPBUILD:41]: Yeah, the version we ship does not have this option.

Comment (carstene1ns) [reply expat/PSPBUILD:41]: Needs 2.2.4 at least, but might be worth to use latest.

Comment (sharkwouter) [reply expat/PSPBUILD:41]: True, but I don't really want to work on that for this PR. It's enough work without upgrades

**Comment by sharkwouter** [comment]
Since MacOS doesn't have the -D option for the ``install`` command, I've updated the current PSPBUILDs to not use it.

**Comment by sharkwouter** [comment]
I've added libpng and libmikmod.

I'm getting closer and closer to only having psp-ports libraries left, though (although I'm not there yet). Do I continue packaging those libraries in the way I did with pspgl or should I wait until they have been moved out of psp-ports?

**Comment by sharkwouter** [comment]
I've added openal, but renamed it to openal-soft. I did add openal to provides.

**Comment by sharkwouter** [comment]
I've added freetype2 from here instead of from psp-ports: https://github.com/pspdev/freetype2

It does rely on a change to psp-pacman which we didn't agree on yet, but I accidentally pushed. Let me know if I need to revert that and add the extra lines to the pspbuild instead: https://github.com/pspdev/psp-pacman/commit/2ece16371e1dd851769d34ffb41443656a6bd830

**Comment by sharkwouter** [comment]
Lua was added thanks to @protheory8

File: pspgl/PSPBUILD (hunk 5 above)
Comment (sharkwouter) [reply pspgl/PSPBUILD]: It now uses the revision of the specific directory within psp-ports instead. So if another package gets upgraded, it does not affect the version of pspgl.

**Comment by sharkwouter** [comment]
Thanks @protheory8 for making PSPBUILDs for libmad, libtremor and libbulletml

**Comment by sharkwouter** [comment]
This should be just about ready now. The resulting libraries should all match what is currently in psplibraries with the exceptions of the SDL/SDL2 based libraries and freetype2 which now use the pspdev repos instead of psp-ports. Fixing problems with the libraries themselves directly, upgrading libraries or adding new libraries was out of scope for this PR.

One possible issue with this PR is that the libraries.sh script can probably not be executed with sudo or as root. I don't think we should change that, though.

File: jpeg/PSPBUILD [hunk 11]
```diff
@@ -0,0 +1,32 @@
+# Maintainer: Wouter Wijsman <wwijsman@live.nl>
+pkgname=jpeg
+pkgver=8c
+pkgrel=1
+pkgdesc="provides read access to zipped files in a zip-archive"
```
Comment (carstene1ns) [inline jpeg/PSPBUILD:5]: no.

File: libmikmod/PSPBUILD [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+prepare() {
+	cd "$pkgname-$pkgver"
+	patch -p1 < ../${pkgname}-${pkgver}-PSP.patch
+}
+
+build() {
+	cd "$pkgname-$pkgver"
+	LDFLAGS="$LDFLAGS -L$(psp-config --pspsdk-path)/lib -L$(psp-config --psp-prefix)/lib -lc -lpspuser" \
+    LIBS="$LIBS -lc -lpspuser" \
```
Comment (carstene1ns) [inline libmikmod/PSPBUILD:29]: whitespace

File: pspgl/PSPBUILD [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	make
+}
+
+package() {
+	cd "psp-ports/$pkgname"
+	# make install isn't usable for pspgl, so we're installing manually
+	mkdir -m 755 -p "$pkgdir/psp/lib" "$pkgdir/psp/include"
+	install -Dm 644 "libGL.a" "$pkgdir/psp/lib/libGL.a"
+	install -Dm 644 "libGLU.a" "$pkgdir/psp/lib/libGLU.a"
+	install -Dm 644 "libglut.a" "$pkgdir/psp/lib/libglut.a"
```
Comment (carstene1ns) [inline pspgl/PSPBUILD:32]: remove `D` and `lib*.a` at the end.

File: pspirkeyb/PSPBUILD [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+pkgver() {
+	cd psp-ports
+	# Set the version to the revision of the pspirkeyb directory
+	printf "r%s" "$(git rev-list --count HEAD ${pkgname})"
+}
+
+build() {
+	cd "psp-ports/$pkgname"
+	make
+	make -C libpspirkeyb
```
Comment (carstene1ns) [inline pspirkeyb/PSPBUILD:24]: why is this needed?

File: squirrel/PSPBUILD [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+	cd "SQUIRREL3"
+	AR="psp-ar" CC="psp-gcc" CXX="psp-g++" LIBS="-lc -lpspuser" make -C squirrel sq32
+	AR="psp-ar" CC="psp-gcc" CXX="psp-g++" LIBS="-lc -lpspuser" make -C sqstdlib sq32
+}
+
+package() {
+	cd "SQUIRREL3"
+
+	mkdir -m 755 -p "$pkgdir/psp/lib"
+	install -m 644 "lib/libsqstdlib.a" "$pkgdir/psp/lib/libsqstdlib.a"
```
Comment (carstene1ns) [inline squirrel/PSPBUILD:36]: `lib*.a` can be removed at the end.

**Comment by carstene1ns** [comment]
Please fix your whitespaces! Decide on one style and stick to it.
- Initial tabs and spaces mixed
- Spaces between command arguments are mixed as well (e.g. `-m755` vs. `-m 755`)

File: pspirkeyb/PSPBUILD (hunk 14 above)
Comment (sharkwouter) [reply pspirkeyb/PSPBUILD]: It wasn't

File: jpeg/PSPBUILD (hunk 11 above)
Comment (sharkwouter) [reply jpeg/PSPBUILD]: Good catch

**Comment by sharkwouter** [comment]
Okay now every whitespace is 4 spaces, every install command uses a space after ``-m`` and no ``install`` or ``cp`` command should now repeat the filename.

File: expat/PSPBUILD [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+pkgdesc="XML parsing C library"
+arch=('mips')
+url="https://libexpat.github.io/"
+license=('MIT')
+depends=()
+makedepends=()
+optdepends=()
+source=(
+    "http://sourceforge.net/projects/expat/files/${pkgname}/${pkgver}/${pkgname}-${pkgver}.tar.gz"
+    "config.sub"
```
Comment (carstene1ns) [inline expat/PSPBUILD:14]: I think we should not really include this config sub everywhere.
Maybe add it as external source file?
`https://raw.githubusercontent.com/pspdev/psplibraries/c4746da120f66f08e07d714334fcf557e259f6cc/patches/config.sub`

File: jpeg/PSPBUILD [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+license=('custom')
+depends=()
+makedepends=()
+optdepends=()
+source=("git+https://github.com/pspdev/psp-ports.git")
+sha256sums=('SKIP')
+
+prepare() {
+    cd psp-ports
+    git checkout f47bd37edf202284fc039106a5b27d9b7314bd76
```
Comment (carstene1ns) [inline jpeg/PSPBUILD:17]: This can be done directly in source array:
`git+https://github.com/pspdev/psp-ports.git#commit=f47bd37edf202284fc039106a5b27d9b7314bd76`

File: libtremor/PSPBUILD [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+sha256sums=('SKIP')
+
+prepare() {
+    cd "psp-ports"
+    git checkout f47bd37edf202284fc039106a5b27d9b7314bd76
+}
+
+build() {
+    cd "psp-ports/libTremor"
+    LDFLAGS="-L$(psp-config --pspsdk-path)/lib" LIBS="-lc -lpspuser" ./autogen.sh --host psp --prefix="/psp" || { exit 1; }
```
Comment (carstene1ns) [inline libtremor/PSPBUILD:22]: Remove `|| …`, this is unneeded.

File: sdl2/PSPBUILD [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    cd SDL2
+    make -f Makefile.psp
+    make -f Makefile.main.psp
+}
+
+package() {
+    cd SDL2
+    mkdir -m 755 -p "$pkgdir/psp/lib" 
+    install -m 644 "libSDL2.a" "$pkgdir/psp/lib/libSDL2.a"
+    install -m 644 "libSDL2main.a" "$pkgdir/psp/lib/libSDL2main.a"
```
Comment (carstene1ns) [inline sdl2/PSPBUILD:31]: Why not `install -m 644 libSDL2.a libSDL2main.a "$pkgdir/psp/lib/"`?

File: sqlite/PSPBUILD [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+prepare() {
+    cd "$pkgname-$pkgver"
+    patch -p1 < ../${pkgname}-${pkgver}-PSP.patch
+}
+
+build() {
+    cd "$pkgname-$pkgver"
+    LDFLAGS="-L$(psp-config --pspsdk-path)/lib -lc -lpspuser" \
+    CFLAGS="-DSQLITE_OS_OTHER=1 -DSQLITE_OS_PSP=1 -I$(psp-config --pspsdk-path)/include" \
+    ./configure --host=psp --prefix="" --disable-readline --disable-tcl --disable-threadsafe --disable-amalgamation
```
Comment (carstene1ns) [inline sqlite/PSPBUILD:30]: Unsetting prefix looks odd, does this work like intended?

File: zlib/PSPBUILD [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+build() {
+    cd "$pkgname-$pkgver"
+    make
+}
+
+package() {
+    cd "$pkgname-$pkgver"
+
```
Comment (carstene1ns) [inline zlib/PSPBUILD:33]: No `make install` available?

**Comment by carstene1ns** [comment]
Sorry, if my comments feel a bit pedantic sometimes. :grinning:

File: sqlite/PSPBUILD (hunk 20 above)
Comment (sharkwouter) [reply sqlite/PSPBUILD]: I have no idea, but I changed it.

File: sdl2/PSPBUILD (hunk 19 above)
Comment (sharkwouter) [reply sdl2/PSPBUILD]: I didn't know that would work, but it did.

File: jpeg/PSPBUILD (hunk 17 above)
Comment (sharkwouter) [reply jpeg/PSPBUILD:17]: This is amazing, I should've known this earlier

Comment (sharkwouter) [reply jpeg/PSPBUILD]: I've done this for all packages which use git now.

File: expat/PSPBUILD (hunk 16 above)
Comment (sharkwouter) [reply expat/PSPBUILD]: Okay, this change removes over 3000 lines from this PR

**Comment by sharkwouter** [comment]
Sorry for the commit spam, I rebased on master to get rid of the conflict

**Comment by sharkwouter** [comment]
This is still missing a change from #71 and the repo change for pspgl.

**Comment by sharkwouter** [comment]
I'm going to close this in favor or creating a new repository.
